### PR TITLE
Feature/458 fi mr scalable

### PR DESCRIPTION
**Pull request by jswaro**
This is the version 1.0 of the FI_MR_SCALABLE implementation. 

Try to focus on big issues for the moment, rather than syntax, formatting, etc... as I'll be adding more commits after this to clean up the PR and add the appropriate documentation. This is a fully functional model based on the requirements from @agontarek and @sdvormwa. 

If you see more tests that should be added, please add comments to that affect. If something doesn't function as you'd expect, let's discuss. 

I will add at least two more commits. One for documentation, and one for clean up. I will send notifications to all reviewers when the final two commits have been added.

As a note, depending on how quickly the working group approves ofi-wg/libfabric#2668 and ofi-wg/libfabric#2729, I may remove the domain extension prior to merge in favor of fi_mr_refresh. 

@sungeunchoi @sdvormwa @chuckfossen @agontarek

### CODE REVIEW

**Comment by hppritcha** [comment]
Does this PR impact performance of apps using '''FI_MR_BASIC'''?

**Comment by hppritcha** [comment]
Also, before merging please squash down to one commit.

**Comment by jswaro** [comment]
> Does this PR impact performance of apps using '''FI_MR_BASIC'''?

I did not notice any impact. What part of the code concerns you? 

> Also, before merging please squash down to one commit.

I was already planning to.

File: man/fi_gni.7.md [hunk 1]
```diff
@@ -22,6 +22,10 @@ support for common atomic operations and optimized collectives.
 The GNI provider runs on Cray XC systems running CLE 5.2 UP04 or higher
 using gcc version 4.9 or higher.
 
+When using the FI_MR_SCALABLE memory mode, applications must make 
+registration requests on 4K page alignment boundaries. Any other value
```
Comment (jswaro) [inline man/fi_gni.7.md]: Change this to 'page boundaries'

File: man/fi_gni.7.md [hunk 2]
```diff
@@ -215,17 +272,12 @@ The `set_val` function sets the value of a given parameter; the
...
-exist in case that they may be reused in the near future. Flushing the stale
-registrations forces hardware-level deregistration of the stale memory
-registrations and frees any memory related to those stale registrations. Only
-the provider-level registration struct is freed, not the user buffer
-associated with the registration.
-The parameter for `flush_cache` is a struct fid pointer to a fi_domain. The
-memory registration cache is tied to the domain, so issuing a `flush_cache` to
-the domain will flush the registration cache of the domain.
+*GNI_USER_WILL_UPDATE_REGIONS*
+: If set, the user is responsible for making the appropriate calls to the
```
Comment (jswaro) [inline man/fi_gni.7.md]: Only 1.5 feature so this isn't needed.

File: man/fi_gni.7.md [hunk 3]
```diff
@@ -254,10 +306,40 @@ but adds the following parameter:
...
+
+The GNI provider sets the default user registration limit to 192 when using FI_MR_SCALABLE,
+and sets the default provider registration limit to 64. These limits are directly associated
+with the authorization key in use when creating the registration. If no authorization key
+is used when creating the registration, the registration is automatically bound to the same
+authorization key as the domain to which the registration belongs. 
+
+When using FI_MR_SCALABLE, the provider may make registrations which consume some of the 
+registrations set aside for the provider. This impacts the performance of FI_LOCAL_MR, which 
+relies on provider-created registrations.
```
Comment (jswaro) [inline man/fi_gni.7.md]: Wording.

File: man/fi_gni.7.md [hunk 4]
```diff
@@ -254,10 +306,40 @@ but adds the following parameter:
...
+
+All memory registrations are associated with an authorization key, whether it is the provider
+default key(keylen=0) or a user-acquired key (key!=NULL, keylen!=0). Each authorization
+key is associated with a unique GNI network key. A GNI network key can only accomodate a single
+memory mode, whether it is FI_MR_BASIC, or FI_MR_SCALABLE. If a user attempts to open multiple 
+domains using different memory modes with the same authorization key, the provider will return
+-FI_EINVAL.
```
Comment (jswaro) [inline man/fi_gni.7.md]: WARN

File: man/fi_gni.7.md [hunk 5]
```diff
@@ -254,10 +306,40 @@ but adds the following parameter:
...
+
+All memory registrations are associated with an authorization key, whether it is the provider
+default key(keylen=0) or a user-acquired key (key!=NULL, keylen!=0). Each authorization
+key is associated with a unique GNI network key. A GNI network key can only accomodate a single
+memory mode, whether it is FI_MR_BASIC, or FI_MR_SCALABLE. If a user attempts to open multiple 
+domains using different memory modes with the same authorization key, the provider will return
+-FI_EINVAL. 
+
+When using FI_MR_SCALABLE, the user may request keys beginning at 0, and ending at the 
+user registration limit for a given authorization key.
```
Comment (jswaro) [inline man/fi_gni.7.md]: This is affected by the global limit of 4 VMDH tables. Fixing this might require doing offset by local rank or something.

Comment (jswaro) [reply man/fi_gni.7.md]: See #1244

File: prov/gni/include/fi_ext_gni.h [hunk 6]
```diff
@@ -77,6 +78,9 @@ typedef enum ep_ops_val {
 #define FI_GNI_FAB_OPS_1 "fab ops 1"
 typedef enum fab_ops_val {
 	GNI_WAIT_THREAD_SLEEP = 0,
+	GNI_DEALLOC_AKI_ON_FABRIC_CLOSE,
```
Comment (jswaro) [inline prov/gni/include/fi_ext_gni.h]: TODO or postpone.

File: man/fi_gni.7.md [hunk 7]
```diff
@@ -254,10 +306,40 @@ but adds the following parameter:
...
+	
+When using FI_MR_SCALABLE and fi_mr_refresh(), only refresh the updated pages, not the entire
+registration. If the entire registration is refreshed and some of the pages are not mapped,
+then refresh will return -FI_EFAULT.
```
Comment (sungeunchoi) [inline man/fi_gni.7.md]: The wording here is awkward.  Maybe something like:

> Only mapped pages can be used with calls to fi_mr_refresh().  If any pages in the range are not mapped, -FI_EFAULT will be returned.

**Comment by sungeunchoi** [comment]
- Make gnitest changes a separate commit
- Add tests to find limits of ugni resource usage and insure that we have reasonable error reporting
- Document limits and return codes

**Comment by sungeunchoi** [comment]
For general testing with gnitest:

- Always run with original MR option (FI_MR_BASIC) for most
- Add tests specifically targeting FI_MR_SCALABLE with expected usage (register all of memory)
-  If there are any existing tests that would be valuable to run in both mode, create a new test/suite so that there's one for FI_MR_BASIC and one for FI_MR_SCALABLE.

For any future changes to FI_MR_SCALABLE, use GNITEST_USE_FI_MR_SCALABLE=1 when running gnitest.

File: prov/gni/src/gnix_fabric.c [hunk 8]
```diff
@@ -660,11 +692,187 @@ __gnix_fab_ops_set_val(struct fid *fid, fab_ops_val_t t, void *val)
...
+		info = _gnix_auth_key_lookup(auth_key, auth_keylen);
+	} else {
+		GNIX_INFO(FI_LOG_FABRIC, "multiple authorization keys not yet"
+			"supported\n");
+		return -FI_ENOSYS; 
+	}
+
+	if (!info) {
+		ret = __gnix_auth_key_initialize(auth_key, auth_keylen, NULL);
+		assert(ret == FI_SUCCESS);
```
Comment (e-harvey) [inline prov/gni/src/gnix_fabric.c]: My understanding is that these asserts are compiled out without debugging enabled. Maybe use `GNIX_FATAL` instead?

File: prov/gni/src/gnix_fabric.c [hunk 9]
```diff
@@ -660,11 +692,187 @@ __gnix_fab_ops_set_val(struct fid *fid, fab_ops_val_t t, void *val)
...
+
+		info = _gnix_auth_key_lookup(auth_key, auth_keylen);
+		assert(info);
```
Comment (e-harvey) [inline prov/gni/src/gnix_fabric.c]: Same as above.

File: prov/gni/src/gnix_init.c [hunk 10]
```diff
@@ -150,6 +151,15 @@ void _gnix_init(void)
 	static int called=0;
```
Comment (e-harvey) [inline prov/gni/src/gnix_init.c]: Does this pass the pre-commit hook?

File: prov/gni/src/gnix_init.c [hunk 11]
```diff
@@ -150,6 +151,15 @@ void _gnix_init(void)
...
 
 	if (called==0) {
```
Comment (e-harvey) [inline prov/gni/src/gnix_init.c]: Same as above.

File: prov/gni/src/gnix_init.c [hunk 12]
```diff
@@ -150,6 +151,15 @@ void _gnix_init(void)
...
+		if (sizeof(struct gnix_mr_key) != 8) {
+			GNIX_FATAL(FI_LOG_FABRIC, 
+				"gnix_mr_key size is invalid, "
+				"size=%d expected=8\n", 
+				sizeof(struct gnix_mr_key));
+			assert(0);
```
Comment (e-harvey) [inline prov/gni/src/gnix_init.c]: Is this reachable?

File: prov/gni/src/gnix_mbox_allocator.c [hunk 13]
```diff
@@ -339,10 +342,34 @@ static int __create_slab(struct gnix_mbox_alloc_handle *handle)
 	}
 
 	COND_ACQUIRE(handle->nic_handle->requires_lock, &handle->nic_handle->lock);
+	if (handle->nic_handle->mr_mode == FI_MR_SCALABLE) {
+		info = _gnix_auth_key_lookup(GNIX_PROV_DEFAULT_AUTH_KEY,
+				GNIX_PROV_DEFAULT_AUTH_KEYLEN);
+		assert(info);
```
Comment (e-harvey) [inline prov/gni/src/gnix_mbox_allocator.c]: Same as above.

File: prov/gni/src/gnix_mbox_allocator.c [hunk 14]
```diff
@@ -339,10 +342,34 @@ static int __create_slab(struct gnix_mbox_alloc_handle *handle)
...
+
+		if (!handle->nic_handle->mdd_resources_set) {
+		/* check to see if the ptag registration limit was set
+		   yet or not -- becomes read-only after success */
+			_gnix_auth_key_enable(info);
+		
+			status = GNI_SetMddResources(
+				handle->nic_handle->gni_nic_hndl,
+				info->attr.prov_key_limit + info->attr.user_key_limit);
+			assert(status == GNI_RC_SUCCESS);
```
Comment (e-harvey) [inline prov/gni/src/gnix_mbox_allocator.c]: Same as above.

File: prov/gni/src/gnix_mbox_allocator.c [hunk 15]
```diff
@@ -339,10 +342,34 @@ static int __create_slab(struct gnix_mbox_alloc_handle *handle)
...
+
+			handle->nic_handle->mdd_resources_set = 1;
+		}
+
+		vmdh_index = _gnix_get_next_reserved_key(info);
+		// TODO: handle error condition
+		assert(vmdh_index > 0);
```
Comment (e-harvey) [inline prov/gni/src/gnix_mbox_allocator.c]: Same as above.

File: prov/gni/src/gnix_mr.c [hunk 16]
```diff
@@ -295,14 +360,101 @@ DIRECT_FN int gnix_mr_regattr(struct fid *fid, const struct fi_mr_attr *attr,
...
+
+
+	fastlock_acquire(&desc->nic->lock);
+	GNIX_INFO(FI_LOG_MR, "hndl=%p addr=%p len=%d cq=%p flags=%x vmdh_index=%d mem_hndl=%p\n",
+		desc->nic->gni_nic_hndl, addr, len, NULL, GNI_MEM_UPDATE_REGION,
+		desc->mr_fid.key, &desc->mem_hndl);
+	grc = GNI_MemRegister(desc->nic->gni_nic_hndl, addr,
+		len, NULL, GNI_MEM_UPDATE_REGION,
+		desc->mr_fid.key, &desc->mem_hndl);
+	assert(grc == FI_SUCCESS); /* TODO: handle error case */
```
Comment (e-harvey) [inline prov/gni/src/gnix_mr.c]: Same as above.

Comment (jswaro) [reply prov/gni/src/gnix_mr.c]: I'd not worry about those yet. Those are in-progress things to remove as part of cleanup. Disregard the asserts until the final review.

File: prov/gni/src/gnix_mr.c [hunk 17]
```diff
@@ -295,14 +360,101 @@ DIRECT_FN int gnix_mr_regattr(struct fid *fid, const struct fi_mr_attr *attr,
...
+	for (i = 0; i < modify->attr.iov_count; i++) {
+		addr = (uint64_t) modify->attr.mr_iov[i].iov_base;
+		len = (uint64_t) modify->attr.mr_iov[i].iov_len;
+
+		if (!(desc->addr <= (uint64_t) addr && (((uint64_t) addr + len) <= (desc->addr + desc->len))))
+			return -FI_EINVAL;
+
+		aligned_addr = addr & ~0xfff;
+		aligned_len = addr + len - aligned_addr;
+		aligned_len += (((addr + len) & 0xfff) ? (0x1000 - ((addr + len) & 0xfff)) : 0);
```
Comment (e-harvey) [inline prov/gni/src/gnix_mr.c]: Could we do `aligned_len = len + (addr - aligned_addr)` instead? Also, is `aligned_addr + aligned_len` supposed to result in a aligned address?

File: prov/gni/src/gnix_mr.c (hunk 16 above)
Comment (e-harvey) [reply prov/gni/src/gnix_mr.c]: Ok, I would be worried about silent failures. Maybe for testing if you're not compiling with --enable-debug you could do something like:
```
#ifndef ENABLE_DEBUG
#define assert(cond) (if ((cond)) {printf("assert failed at %d\n", __LINE__); abort();})
#endif
```

File: prov/gni/test/fabric.c [hunk 18]
```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright (c) 2015-2016 Cray Inc. All rights reserved.
```
Comment (e-harvey) [inline prov/gni/test/fabric.c]: 2017?

File: prov/gni/test/fabric.c [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+	cr_assert(fabric != NULL);
+}
+
+Test(fabric, open_ops_1)
+{
+	int ret;
+	struct fi_gni_ops_fab *ops; 
+
+	ret = fi_open_ops(&fabric->fid, FI_GNI_FAB_OPS_1, 0, (void **) &ops, NULL);
+	cr_assert_eq(ret, FI_SUCCESS);
```
Comment (e-harvey) [inline prov/gni/test/fabric.c]: I found it useful to use: `cr_assert(ret == FI_SUCCESS, "fi_open_ops returned: %s", fi_strerror(-ret));` while testing the FI_ADDR_STR changes.

**Pull request closed without merging**